### DISCUSSION TITLE
Rendering non-Livewire pagination after rendering Livewire pagination does not work when running on Octane

### DISCUSSION THREAD

**Discussion by dejagersh** [livewire/livewire#5351] [7. Bugs (archived)] [answered]
EDIT:
Problem: If using Octane you first view a page that has Livewire pagination, and then you use non-Livewire pagination, it breaks. This is because the paginator view does not get reset between requests.

**Original post before I figured out the problem:**

I have an endpoint (let's call it `/reviews`) that loads a view with "reviews" as such:

```php
class ReviewController extends Controller
{
    public function __invoke(): View
    {
        $website = $this->getWebsite();
        $team = $website->team;

        return view('sites.reviews', [
            'website' => $website,
            'team' => $team,
            'reviews' => $team->reviews()->orderByDesc('published_at')->paginate(10),
        ]);
    }
}
```

In that view file I have the following code:

```blade
<div>
    @foreach($reviews as $review)
        <!-- Omitted for brevity -->
    @endforeach

    @if($reviews->hasPages())
        <div
            class="py-2 px-4 2xl:mb-0 mb-16 bg-white shadow-sm rounded-b-md">
            {{ $reviews->links() }}
        </div>
    @endif
</div>
```

This page does not load Livewire using `@livewireScripts` or `@livewireStyles`.

This works fine until it hits production. Then it _sometimes_ fails on the following error:

```Using $this when not in object context (View: /var/www/html/vendor/livewire/livewire/src/views/pagination/tailwind.blade.php```

After a couple of minutes it works again. Very odd. I have one interesting finding:

I have an internal dashboard that shows the reviews using Livewire (let's call it `/dashboard/reviews`). It's basically the same list, except it uses Livewire. As mentioned before, after I get the 500 error on `/reviews`, it works again after a couple of minutes. If I _then_ visit the `/dashboard/reviews` page it breaks `/reviews`.

It feels like a bug in how cached views are resolved. My non-Livewire view file is loading in a cached Livewire view file.

Appendix A: full error
<img width="1060" alt="image" src="https://user-images.githubusercontent.com/9981548/200582382-33a1e78c-12bd-4728-a81b-2cd3e063e435.png">

**Comment by dejagersh** [comment]
I think I figured it out:

```php
Paginator::defaultView($this->paginationView());
Paginator::defaultSimpleView($this->paginationSimpleView());
```

This is called in the `WithPagination` trait. These functions set a `static` variable on the `AbstractPaginator` class. It configures it to use the Livewire pagination view. I am using Octane so this variable isn't reverted. Subsequent calls to `{{ $reviews->render() }}` will use the Livewire paginator.

This also explains why it works again after a certain amount of time. I have the Octane process configured to handle a certain amount of requests before it is restarted.

**Comment by dejagersh** [comment, marked as answer]
Josh Hanley told me this is a known issue 👍 For anyone running into the same issue: manually resetting it back to the correct value before rendering any non-Livewire paginators is an easy fix until this gets fixed in Livewire. I've done this using a listener that responds to the `Laravel\Octane\Events\RequestReceived` event:

```php
class ResetPaginator
{
    public function handle(): void
    {
        Paginator::useTailwind();
    }
}
```

You can register it by specifying it in `config/octane.php`: 

```diff
RequestReceived::class => [
    ...Octane::prepareApplicationForNextOperation(),
    ...Octane::prepareApplicationForNextRequest(),

+    ResetPaginator::class
],
```

**Reply by joshhanley** [reply]
@dejagersh sweet! Thanks for sharing 🙂
